src/mod_muc_light_utils.erl

```diff
@@ -246,90 +240,113 @@ value2b(Val, float) -> float_to_binary(Val).
 
 %% ---------------- Affiliations manipulation ----------------
 
--spec fix_edge_aff_users_case(ChangeResult :: change_aff_success() | {error, bad_request}) ->
+-spec maybe_select_new_owner(ChangeResult :: change_aff_success() | {error, bad_request}) ->
     change_aff_success() | {error, bad_request}.
-fix_edge_aff_users_case({ok, NewAffUsers0, AffUsersChanged0, JoiningUsers, LeavingUsers}) ->
-    {NewAffUsers, AffUsersChanged} =
-    case NewAffUsers0 of
-        [{Owner, member}] -> % edge case but can happen
-            {[{Owner, owner}],
-             case lists:member(Owner, JoiningUsers) of
-                 false ->
-                     %% Owner is an old owner
-                     lists:keydelete(Owner, 1, AffUsersChanged0);
-                 true ->
-                     %% Owner is a newcomer, should be promoted
-                     lists:keyreplace(Owner, 1, AffUsersChanged0, {Owner, owner})
-             end};
-        _ ->
-            {NewAffUsers0, AffUsersChanged0}
-    end,
-    {ok, NewAffUsers, AffUsersChanged, JoiningUsers, LeavingUsers};
-fix_edge_aff_users_case({error, bad_request}) ->
-    {error, bad_request}.
+maybe_select_new_owner({ok, AU, AUC, JoiningUsers, LeavingUsers} = AffRes) ->
+    {AffUsers, AffUsersChanged} =
+        case {lists:keyfind(owner, 2, AU), find_new_owner(AffRes)} of
+            {false, {NewOwner, PromotionType}} -> %select new owner
+                NewAU = lists:keyreplace(NewOwner, 1, AU, {NewOwner, owner}),
+                NewAUC = case PromotionType of
+                             promote_old_member ->
+                                 [{NewOwner, owner} | AUC];
+                             promote_joined_member ->
+                                 lists:keyreplace(NewOwner, 1, AUC, {NewOwner, owner});
+                             promote_demoted_owner ->
+                                 lists:keydelete(NewOwner, 1, AUC)
+                         end,
+                {NewAU, NewAUC};
+
+            _ ->
+                {AU, AUC}
+        end,
+    {ok, AffUsers, AffUsersChanged, JoiningUsers, LeavingUsers};
+maybe_select_new_owner(Error) ->
+    Error.
+
+-spec find_new_owner(ChangeResult :: change_aff_success()) ->
+    {jid:simple_bare_jid(), promotion_type()} | false.
+find_new_owner({ok, AU, AUC, JoiningUsers, _LeavingUsers}) ->
+    AllMembers = [U || {U, member} <- (AU)],
+    NewMembers = [U || {U, member} <- (AUC)],
+    OldMembers = AllMembers -- NewMembers,
+    DemotedOwners = NewMembers -- JoiningUsers,
+    %% try to select the new owner from:
+    %%   1) old unchanged room members
+    %%   2) new just joined room members
+    %%   3) demoted room owners
+    case {OldMembers, JoiningUsers, DemotedOwners} of
```

Comment:
Would it be a good idea to put it to a seperate function so that this big, complicated case is named? Like `elect_new_owner`?

Reply:
or rather `promote_new_owner`